### PR TITLE
Bluetooth: GATT: Fix bt_gatt_discovery with BT_GATT_DISCOVER_DESCRIPTOR

### LINKED ISSUE #16602: Bluetooth: GATT Discovery: Descriptor Discovery by range Seg Fault
**Describe the bug**
Descriptor discovery by range ends in seg fault.

**To Reproduce**
1. Modify the discovery code in _central_hr_ sample to 
```		
discover_params.uuid = NULL;
discover_params.func = discover_func;
discover_params.start_handle = 0x0001;
discover_params.end_handle = 0xffff;
discover_params.type = BT_GATT_DISCOVER_DESCRIPTOR;
```

**Expected behavior**
Discovery should work.

**Impact**
Showstopper.

**Environment (please complete the following information):**
 - OS: (e.g. Linux, MacOS, Windows): Kubuntu 18.04
 - Toolchain: Zephyr SDK 0.10.0
 - Commit SHA or Version used: 2692f2e616c7f0bfcd19fda8010ab0c124f38cf5

### PR DESCRIPTION
**Pull request by Vudentz**
Fix iterating past the response which causes an invalid memory to be
accessed and passed over to the callback as if there were more
attributes found.
    
Fixes #16602

Note: Only the first patch need to be backported, the second was used just for checking if things are working properly.

### CODE REVIEW

**Comment by Vudentz** [comment]
@paulgavrikov let me know if this doesn't solve your problem.

**Comment by paulgavrikov** [comment]
@Vudentz doesn't seem to fix the issue.

**Comment by Vudentz** [comment]
> @Vudentz doesn't seem to fix the issue.

With the shell it seems to be working properly which something pretty similar to the code you had, are you sure the code is not trying to reuse the params?

**Comment by Vudentz** [comment]
@paulgavrikov I just tried shell with a nrf52 board:

uart:~$ gatt discover-descriptor 
Discover pending
Descriptor 2902 found: handle 9
Discover complete

It is not reproducible for me anymore.

**Comment by paulgavrikov** [comment]
@Vudentz  are you returning `BT_GATT_ITER_CONTINUE` if attr is not null?

**Comment by Vudentz** [comment]
> @Vudentz are you returning `BT_GATT_ITER_CONTINUE` if attr is not null?

Nope, it return BT_GATT_ITER_STOP:

https://github.com/zephyrproject-rtos/zephyr/blob/master/subsys/bluetooth/shell/gatt.c#L116

I thought you were also stopping or you are not?

**Comment by Vudentz** [comment]
@paulgavrikov try again, it seems the condition in which we check for characteristic and skip the value was causing the iterator to loop around.

**Comment by Vudentz** [comment]
@paulgavrikov any chance to give the lastest version a try?

**Comment by Vudentz** [comment]
@jhedberg then perhaps you can have a look.

File: subsys/bluetooth/host/gatt.c [hunk 1]
```diff
@@ -2470,10 +2471,16 @@ static void gatt_find_info_rsp(struct bt_conn *conn, u8_t err,
...
-				if (length >= len) {
-					pdu = (const u8_t *)pdu + len;
-					length -= len;
+				i--;
+				/* If value is the last one on the response
+				 * exit the loop and adjust the handle.
+				 */
+				if (i == 1) {
+					i--;
+					handle += 1;
```
Comment (jhedberg) [inline subsys/bluetooth/host/gatt.c]: How come you're mixing two different increment/decrement conventions on consecutive lines? Why not just handle++?

**Comment by jhedberg** [comment]
> @jhedberg then perhaps you can have a look.

@Vudentz looks ok, except for the small comment I had. Either way I was also waiting for @paulgavrikov to review/test this.

**Comment by paulgavrikov** [comment]
@Vudentz @jhedberg Can't test this before Monday.

File: subsys/bluetooth/host/gatt.c (hunk 1 above)
Comment (Vudentz) [reply subsys/bluetooth/host/gatt.c]: I forgot push the lastest version so the code no longer tries to skip in place and treat the corner case where it is the last attribute, instead Im now just using a flag to tell if the next handle needs to be skipped which is more in line with the spec allowing gaps so we don't make the assumption it is handle++ but parse the handle and only then skip.

**Comment by Vudentz** [comment]
@paulgavrikov let me know when you are done testing this.

**Comment by paulgavrikov** [comment]
@Vudentz @jhedberg Works now. Thanks a lot!

**Review by jhedberg — APPROVED** [review]

**Pull request merged**